container/container.go

```diff
@@ -2383,3 +2403,133 @@ func hasValidBindPermissions(binds []string) error {
+						bExists = true
+						break
+					}
+				}
+			}
+
+			if !bExists {
+				// create the volume if it does not exist
+				vOption := docker.CreateVolumeOptions{
+					Name:   vol_name,
+					Driver: "local",
+					Labels: map[string]string{
+						LABEL_PREFIX + ".service_name": serviceName,
+						LABEL_PREFIX + ".agreement_id": agreementId,
+						LABEL_PREFIX + ".owner":        "openhorizon"},
+				}
+
+				if _, err := b.client.CreateVolume(vOption); err != nil {
+					return fmt.Errorf("Failed to create the docker volume %v for service %v. %v", vol_name, serviceName, err)
+				} else {
+					glog.V(3).Infof("Volume %v created for service %v.", vol_name, serviceName)
+
+					// same the volume in local db so that it can be cleaned up at unregistration time
+					// Ling todo - only save the ones that are specified by the user in binds.
+					if persistence.SaveContainerVolumeByName(b.db, vol_name); err != nil {
+						return fmt.Errorf("Failed to get save the docker volume name %v into the local db. %v", vol_name, err)
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// Delete the docker volumes that are created by anax
+func DeleteLeftoverDockerVolumes(db *bolt.DB, config *config.HorizonConfig) error {
+	glog.V(3).Infof("Cleaning up leftover docker volumes created by anax.")
+
+	// get leftover volume names from local db
+	cvs, err := persistence.FindAllUndeletedContainerVolumes(db)
+	if err != nil {
+		return fmt.Errorf("Error retrieving undeleted container volumes from local db. %v", err)
+	} else if cvs == nil || len(cvs) == 0 {
+		// nothing to remove
+		glog.V(3).Infof("The cleanup process found no leftover docker volumes in the local db.")
+		return nil
+	}
+
+	if client, err := docker.NewClient(config.Edge.DockerEndpoint); err != nil {
+		return fmt.Errorf("Failed to instantiate docker Client: %v", err)
+	} else {
+		// check existing docker volumes
+		volumes_docker, err := client.ListVolumes(docker.ListVolumesOptions{})
+		if err != nil {
+			return fmt.Errorf("Failed to get the docker volumes. %v", err)
+		} else if volumes_docker == nil || len(volumes_docker) == 0 {
+			// no docker volumes
+			glog.V(3).Infof("The cleanup process found no docker volumes.")
+			return nil
+		}
+
+		for _, cv := range cvs {
+
+			// make sure the volume still exists and it has openhorizon as the owner
+			found := false
+			for _, dv := range volumes_docker {
+				if cv.Name == dv.Name {
+					if dv.Labels != nil && dv.Labels[LABEL_PREFIX+".owner"] == "openhorizon" {
+						found = true
+						break
+					}
+				}
+			}
+
+			// remove the volume and remove it from the local db
+			if found {
+				if err := client.RemoveVolume(cv.Name); err != nil {
+					// failure to delete the volume should not prevent the process from going on
+					glog.Errorf("Container sync resources. Failed to delete docker volume %v. %v", cv.Name, err)
+				} else if err := persistence.ArchiveContainerVolumes(db, &cv); err != nil {
```

Comment:
The remove failure is often because the volume is used by other containers at the moment. This is the case for multiple anax instances running on the same host.  The reason we do not remove it from the local db here is that it will have a chance to get cleaned when this anax starts again.